pypdf/_writer.py

```diff
@@ -847,6 +848,46 @@ def _update_text_field(self, field: DictionaryObject) -> None:
             da = " ".join(font_properties)
         y_offset = rct.height - 1 - font_height
 
+        # Retrieve font information from local DR ...
+        dr: Any = cast(
+            DictionaryObject,
+            cast(DictionaryObject, field.get("/DR", DictionaryObject())).get_object(),
+        )
+        dr = dr.get("/Font", DictionaryObject()).get_object()
+        if font_name not in dr:
+            # ...or AcroForm dictionary
+            dr = cast(
+                dict,
+                cast(DictionaryObject, self._root_object["/AcroForm"]).get("/DR", {}),
+            )
+            if isinstance(dr, IndirectObject):
+                dr = dr.get_object()
+            dr = dr.get("/Font", DictionaryObject()).get_object()
+        font_res = dr.get(font_name)
+        if font_res is not None:
+            font_res = cast(DictionaryObject, font_res.get_object())
+            font_subtype, _, font_encoding, font_map = build_char_map_from_dict(
+                200, font_res
+            )
+            font_full_rev: Dict[str, int]
+            if isinstance(font_encoding, str):
+                assert font_encoding in ("charmap", "utf-16-be")
+                if font_encoding not in ("charmap", "utf-16-be"):
+                    logger_warning(
+                        f"unexpected {font_encoding} : please share pdf with pypdf dev team",
+                        __name__,
+                    )
+                font_full_rev = {v: k for k, v in font_map.items()}
+            else:
+                font_full_rev = {v: k for k, v in font_encoding.items()}
+                font_encoding_rev = {v: k for k, v in font_encoding.items()}
+                for k, v in font_map.items():
+                    font_full_rev[v] = font_encoding_rev.get(k, ord(k))
+        else:
+            raise AssertionError("can not find font dictionary")
+            logger_warning(f"can not find font dictionary for {font_name}", __name__)
+            font_full_rev = {}
```

Comment:
The code after the exception will never be executed

I would use ValueError instead of AssertionError: https://docs.python.org/3/library/exceptions.html#exception-hierarchy - but I cannot give a good reason why :thinking:

Reply:
I agree that using *AssertionError*s usually is a bad approach. Why raising an exception at all when there apparently is a way to continue afterwards after having logged a (recoverable) violation/issue?

Reply:
The assertion error is here to detect issues when using pytest. In operational condition it is the logger_warning (non blocking) that should be raised

Reply:
I don't understand that. The Exception will also be raised in prod code, not only in pytest.

When you execute the following with python:

```
print("First")
raise AssertionError("the exception")
print("Second")
```

you get:

```
First
Traceback (most recent call last):
  File "/home/moose/foo.py", line 2, in <module>
    raise AssertionError("the exception")
AssertionError: the exception
```

Reply:
I see my mistake. I originally used assert but ruff changed to the exception only.
I propose this code:
```
if __debug__:
            raise AssertionError("can not find font dictionary")
else:
            logger_warning(f"can not find font dictionary for {font_name}", __name__)
```

your opinion ?

Reply:
Interesting, I didn't know about [`__debug__`](https://docs.python.org/3/library/constants.html#debug__) - thanks for sharing :pray: 

Sadly, I think it cannot be used like that. I think more than 90% of all users don't use/know the `-O` flag.

Reply:
Within pytest, we can make warnings which are not captured via `with pytest.warns` be treated as errors with `pytest -W error`: https://docs.python.org/3/using/cmdline.html#cmdoption-W

Reply:
I think that is probably a good idea as default as well :thinking:

Reply:
within pyproject.toml there is already:
```
[tool.pytest.ini_options]
addopts = "--disable-socket"
filterwarnings = ["error"]
```
isn't this sufficient consider warnings as errors ?
for the moment I removed the raise line

Reply:
Yes, I think that should do it. Didn't it work?